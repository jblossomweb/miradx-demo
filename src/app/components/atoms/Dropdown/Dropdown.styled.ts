import tw from 'tailwind-styled-components'

export interface ContainerProps {
  $disabled?: boolean
}

export interface LabelProps {
  $required?: boolean
}

export interface SelectProps {
  $fullWidth?: boolean
  $error?: boolean
}

export interface OptionProps {
  selected?: boolean
  disabled?: boolean
}

export const Container = tw.label<ContainerProps>`
  flex
  flex-col
  items-left
  font-medium
  ${({ $disabled }) => ($disabled ? 'text-gray-400' : '')}
`

export const Label = tw.span<LabelProps>`
  after:ml-1
  after:text-red-500
  ${({ $required }) => ($required ? `after:content-['*']` : '')}
`

export const Select = tw.select<SelectProps>`
  ${({ $fullWidth }) => ($fullWidth ? 'w-full' : 'w-80')}
  h-8
  text-black
  overflow-visible
  rounded
  ring-1
  focus:ring-blue-500
  ${({ $error }) => ($error ? 'ring-red-500' : 'ring-transparent')}
  ${({ disabled }) => (disabled ? 'cursor-not-allowed' : 'cursor-pointer')}
`

export const Option = tw.option<OptionProps>`
  ${({ disabled }) => (disabled ? 'text-gray-500' : '')}
`

export const Error = tw.span`
  text-red-500
  font-bold
  italic
`
