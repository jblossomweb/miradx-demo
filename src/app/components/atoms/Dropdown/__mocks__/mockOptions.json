[
  { "display": "Option One", "value": "One" },
  { "display": "Option Two", "value": "Two" },
  { "display": "Option Three", "value": "Three" },
  { "display": "Option Four", "value": "Four" }
]
