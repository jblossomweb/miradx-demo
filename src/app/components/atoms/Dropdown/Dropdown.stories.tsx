import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import mockOptions from './__mocks__/mockOptions.json'

import Dropdown, { type DropdownProps } from './Dropdown'

type Story = StoryObj<typeof meta>

const meta = {
  title: 'components/atoms/Dropdown',
  component: Dropdown,
  render: (args) => <Dropdown {...args} />,
} satisfies Meta<typeof Dropdown>

export const Sample: Story = {
  args: {
    label: 'Sample Dropdown',
    options: mockOptions as DropdownProps['options'],
  },
}

export const Disabled: Story = {
  args: {
    label: 'Disabled Dropdown',
    disabled: true,
    options: mockOptions as DropdownProps['options'],
  },
}

export const Required: Story = {
  args: {
    label: 'Required Dropdown',
    required: true,
    options: mockOptions as DropdownProps['options'],
  },
}

export const Placeholder: Story = {
  args: {
    label: 'Sample Dropdown',
    placeholder: 'Sample Placeholder',
    options: mockOptions as DropdownProps['options'],
  },
}

export const NoLabel: Story = {
  args: {
    options: mockOptions as DropdownProps['options'],
  },
}

export const NoLabelRequired: Story = {
  args: {
    required: true,
    options: mockOptions as DropdownProps['options'],
  },
}

export const FullWidth: Story = {
  args: {
    label: 'Full Width Dropdown',
    fullWidth: true,
    options: mockOptions as DropdownProps['options'],
  },
}

export const Controlled: Story = {
  args: {
    label: 'Controlled Dropdown',
    value: mockOptions[0].value,
    onChange: (event) => {
      console.log(`set value to ${event.target.value}`)
    },
    options: mockOptions as DropdownProps['options'],
  },
}

export const WithError: Story = {
  args: {
    label: 'Dropdown With Error',
    error: 'Something is wrong with your selection',
    options: mockOptions as DropdownProps['options'],
  },
}

export default meta
