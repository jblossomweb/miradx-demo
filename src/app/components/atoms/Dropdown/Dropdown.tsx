import React from 'react'
import * as Styled from './Dropdown.styled'

export interface SelectOption {
  display: string
  value: string | number
}

export interface DropdownProps {
  label?: string
  disabled?: boolean
  placeholder?: string
  required?: boolean
  password?: boolean
  fullWidth?: boolean
  options: SelectOption[]
  value?: string | number
  error?: string
  onFocus?: (event: React.FocusEvent<HTMLSelectElement>) => void
  onChange?: (event: React.ChangeEvent<HTMLSelectElement>) => void
}

const Dropdown: React.FC<DropdownProps> = ({
  label,
  disabled,
  placeholder,
  required,
  fullWidth,
  options,
  value,
  error,
  onFocus,
  onChange,
}) => (
  <Styled.Container $disabled={disabled}>
    {!!(label || required) && (
      <Styled.Label $required={required}>{label}</Styled.Label>
    )}
    <Styled.Select
      required={required}
      disabled={disabled}
      aria-label={label}
      defaultValue={value || ''}
      onFocus={onFocus}
      onChange={onChange}
      $fullWidth={fullWidth}
      $error={!!error}
    >
      {!!placeholder && (
        <Styled.Option
          value=""
          disabled
        >
          {placeholder}
        </Styled.Option>
      )}
      {options.map((option) => (
        <Styled.Option
          key={option.value}
          value={option.value}
        >
          {option.display}
        </Styled.Option>
      ))}
    </Styled.Select>
    {!!error && <Styled.Error>{error}</Styled.Error>}
  </Styled.Container>
)

export default Dropdown
