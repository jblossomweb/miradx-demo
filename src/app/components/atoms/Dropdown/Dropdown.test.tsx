import React from 'react'
import { render, screen, within, fireEvent } from '@testing-library/react'
import storySnaps from 'test/utils/storySnaps'
import randomInteger from 'test/generate/randomInteger'

import Dropdown, { type DropdownProps } from './Dropdown'
import * as stories from './Dropdown.stories'
// eslint-disable-next-line jest/no-mocks-import
import mockOptions from './__mocks__/mockOptions.json'

describe('app/components/atoms/Dropdown', () => {
  describe('stories', () => {
    storySnaps(stories)
  })

  describe('assertions', () => {
    const onChange = jest.fn()
    const renderDefault = (options: DropdownProps['options']) =>
      render(<Dropdown options={options} />)

    const renderRequired = (options: DropdownProps['options']) =>
      render(
        <Dropdown
          options={options}
          required
        />,
      )

    const renderLabel = (label: string, options: DropdownProps['options']) =>
      render(
        <Dropdown
          label={label}
          options={options}
        />,
      )

    const renderError = (error: string, options: DropdownProps['options']) =>
      render(
        <Dropdown
          error={error}
          options={options}
        />,
      )

    const renderControlled = (
      selectedIndex: number,
      options: DropdownProps['options'],
    ) =>
      render(
        <Dropdown
          value={options[selectedIndex].value}
          options={options}
          onChange={onChange}
        />,
      )

    beforeEach(jest.clearAllMocks)

    it('renders a Dropdown', () => {
      renderDefault(mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      expect(dropdown).toBeInTheDocument()
    })

    it('renders a Dropdown with specified options', () => {
      renderDefault(mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      const options = within(dropdown).getAllByRole('option')
      expect(options.length).toEqual(mockOptions.length)
      mockOptions.forEach(({ display }, i) => {
        const option = within(dropdown).getByText(display)
        const value = option.getAttribute('value')
        expect(option).toBeInTheDocument()
        expect(value).toEqual(options[i].getAttribute('value'))
      })
    })

    it('indicates to browser when Dropdown selection is required', () => {
      renderRequired(mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      expect(dropdown.required).toBe(true)
    })

    it('renders a label when specified', () => {
      const labelText = 'Test Label'
      renderLabel(labelText, mockOptions)
      const label = screen.getByText(labelText)
      expect(label).toBeInTheDocument()
    })

    it('indicates label to screen readers when specified', () => {
      const labelText = 'Test Label'
      renderLabel(labelText, mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      expect(dropdown.getAttribute('aria-label')).toBe(labelText)
    })

    it('renders an error when specified', () => {
      const errorText = 'Test Error'
      renderError(errorText, mockOptions)
      const error = screen.getByText(errorText)
      expect(error).toBeInTheDocument()
    })

    it('controls selected value when specified', () => {
      const selectedIndex = randomInteger(0, mockOptions.length - 1) // fuzzy test
      const valueText = mockOptions[selectedIndex].value
      renderControlled(selectedIndex, mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      expect(dropdown.value).toBe(valueText)
    })

    it('calls onChange when value is changed', () => {
      const initialValue = mockOptions[0].value
      renderControlled(0, mockOptions)
      const dropdown = screen.getByRole('combobox') as HTMLSelectElement
      expect(dropdown.value).toBe(initialValue)
      expect(onChange).not.toHaveBeenCalled()
      fireEvent.change(dropdown, { target: { value: mockOptions[1].value } })
      expect(onChange).toHaveBeenCalled()
    })
  })
})
