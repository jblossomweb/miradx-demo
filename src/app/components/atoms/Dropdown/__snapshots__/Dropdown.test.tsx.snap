// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/atoms/Dropdown stories Controlled matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Controlled Dropdown
    </span>
    <select
      aria-label="Controlled Dropdown"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
    >
      <option
        class=""
        selected=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories Disabled matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium text-gray-400"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Disabled Dropdown
    </span>
    <select
      aria-label="Disabled Dropdown"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-not-allowed"
      disabled=""
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories FullWidth matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Full Width Dropdown
    </span>
    <select
      aria-label="Full Width Dropdown"
      class="w-full h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories NoLabel matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <select
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories NoLabelRequired matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500 after:content-['*']"
    />
    <select
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
      required=""
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories Placeholder matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Sample Dropdown
    </span>
    <select
      aria-label="Sample Dropdown"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
    >
      <option
        class="text-gray-500"
        disabled=""
        selected=""
        value=""
      >
        Sample Placeholder
      </option>
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories Required matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500 after:content-['*']"
    >
      Required Dropdown
    </span>
    <select
      aria-label="Required Dropdown"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
      required=""
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories Sample matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Sample Dropdown
    </span>
    <select
      aria-label="Sample Dropdown"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-transparent cursor-pointer"
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/Dropdown stories WithError matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Dropdown With Error
    </span>
    <select
      aria-label="Dropdown With Error"
      class="w-80 h-8 text-black overflow-visible rounded ring-1 focus:ring-blue-500 ring-red-500 cursor-pointer"
    >
      <option
        class=""
        value="One"
      >
        Option One
      </option>
      <option
        class=""
        value="Two"
      >
        Option Two
      </option>
      <option
        class=""
        value="Three"
      >
        Option Three
      </option>
      <option
        class=""
        value="Four"
      >
        Option Four
      </option>
    </select>
    <span
      class="text-red-500 font-bold italic"
    >
      Something is wrong with your selection
    </span>
  </label>
</DocumentFragment>
`;
